### PR TITLE
 [Latest Nov 12] Optimize increment summations

### PR DESCRIPTION
**Pull request by laithsakka**
Summary:
**wins**
on torchrec benchmark, for 2K nodes it save 40seconds
with the recent sympy changes (https://www.internalfb.com/diff/D65883538) we save around 13 second. 
```
buck2 run fbcode//mode/opt fbcode//torchrec/distributed/tests:pt2_compile_benchmark -- --num-features=200
```
This diff optimizes construction expressions of the form 
a+b+c...  (all unique symbols). 
which are very common in torchrec models. 

**How**
Expressions of the form a+b+c are not optimized by add, the only needed optimization is sorting them.
If we have  a+b+c and we are adding (d) to it, we can do a binary search to know 
the position of (d) and avoid optimizing the new expression by passing the new order. 

the key here is
1.  avoiding the compare N times and to it Log(N) times.
2. avoid other opts in Add that are not needed.
3. 
Screenshot 2024-11-13 at 11 06 57 PM

**Extensions**:
1. support constant terms.
2. support 10a+10b+.. (this will give even more wins will extend the support in second PR)

Test Plan: add tests in the diff and run exisiting tests.

Differential Revision: D65837755

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

### CODE REVIEW

File: torch/fx/experimental/sym_node.py [hunk 1]
```diff
@@ -755,8 +755,83 @@ def _sympy_rshift(a, b):
...
+    while low <= high:
+        mid = (low + high) // 2
+        if ordered_args[mid] == new_arg:
+            return None
+        elif sort_key(ordered_args[mid]) < sort_key(new_arg):
+            low = mid + 1
+        else:
+            high = mid - 1
+
+    ordered_args.insert(low, new_arg)
```
Comment (ezyang) [inline torch/fx/experimental/sym_node.py:779]: Doesn't this make this function O(n) anyways though??

Comment (laithsakka) [reply torch/fx/experimental/sym_node.py:779]: oh yeh good catch but this is saves bunch of seconds it reduces the comparisons. 
just ran it again on 1000 nodes. and confirmed wins evey with recent sympy changes. 
maybe the add does more than just the sorting?

Comment (laithsakka) [reply torch/fx/experimental/sym_node.py:779]: one thing is that comparing symbols is expensive, I know the recent sympy changes optimizes that by caching stuff.

Comment (laithsakka) [reply torch/fx/experimental/sym_node.py:779]: here we go that answers it 
Screenshot 2024-11-13 at 11 06 57 PM

the profile of constructing Add

Comment (ezyang) [reply torch/fx/experimental/sym_node.py:779]: sg

**Comment by laithsakka** [comment]
this is not being sycned with the diff i will have to close it and open a new PR 
!!!!!!!

**Pull request closed without merging**